File: Leetcode/538-Convert-BST-To-Greater-Tree/ConvertBSTToGreaterTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var convertBST = function(root) {
    if (!root) {
        return root
    }
    var convert = function(node, sum) {
        if (!node) {
            return sum
        }
        node.val += convert(node.right, sum)
        return convert(node.left, node.val)
    };
    convert(root, 0)
    return root
};

File: Leetcode/538-Convert-BST-To-Greater-Tree/ConvertBSTToGreaterTree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func convertBST(root *TreeNode) *TreeNode {
	if root == nil {
		return root
	}
	convert(root, 0)
	return root
}

func convert(node *TreeNode, sum int) int {
	if node == nil {
		return sum
	}
	node.Val += convert(node.Right, sum)
	return convert(node.Left, node.Val)
}
